### PR TITLE
Optimize basetype caches

### PR DESCRIPTION
**Pull request by odersky**
Main optimization: Invalidate base type caches only if runId changes

Phase-ids should not change base type caches. Instead we make sure
that class denotations that contain base type caches are copied
whenever base types might change.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Review by darkknight100 — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 1]
```diff
@@ -2641,7 +2641,11 @@ object SymDenotations {
     private var locked = false
     private var provisional = false
 
-    final def isValid(using Context): Boolean = valid && isValidAt(ctx.phase)
+    final def isValid(using Context): Boolean =
+      valid && createdAt.runId == ctx.runId
+        // Note: We rely on the fact that whenever base types of classes change,
+        // the affected classes will get new denotations with new basedata caches.
+        // So basedata caches can become invalid only if the run changes.
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:2648]: I'm not familiar with code in this area. I'm wondering will the cache be always valid for classes defined in libraries,  regardless of runs?

**Comment by odersky** [comment]
We get a ~5% improvement for dotc and a 8% improvement for re2. No change for scalap and 1-2% worse for stdlib. That could be in the noise though. Still I wonder why stdlib does not do better here. Maybe a low cache hit rate for base type caches? That would make sense as we have lots of small methods with their own type parameters. With a low hit rate, maybe fewer cache invalidations (which is what this PR achieves) means larger caches that are under-utilized. But that's speculation.

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/SymDenotations.scala:2648]: Yes, but libraries might be redefined in source in the next run, so you need to re-check.

**Pull request merged**